Px.Utils/Validation/DataValidation/DataValidator.cs

```diff
@@ -0,0 +1,218 @@
+        /// Assumes that the stream is at the start of the data section (after 'DATA='-keyword) at the first data item.
+        /// </summary>
+        /// <param name="stream">The stream containing the data to validate.</param>
+        /// <param name="rowLen">The expected length of each row in the data.</param>
+        /// <param name="numOfRows">The expected number of rows in the data.</param>
+        /// <param name="startRow">The starting row number.</param>
+        /// <param name="streamEncoding">The encoding of the stream. Can be null, in which case the default encoding is used.</param>
+        /// <param name="conf">The configuration for validating the data file syntax. Can be null, in which case the default configuration is used.</param>
+        /// <param name="cancellationToken">A cancellation token that can be used to cancel the operation.</param>
+        /// <returns>A collection of validation feedback for the data.</returns>
+        public async Task<IEnumerable<ValidationFeedback>> ValidateAsync(Stream stream, int rowLen,
+            int numOfRows,
+            int startRow, Encoding? streamEncoding, PxFileSyntaxConf? conf = null,
+            CancellationToken cancellationToken = default)
+        {
+            SetValidationParameters(streamEncoding, conf, numOfRows, rowLen, startRow);
+
+            IEnumerable<ValidationFeedback> validationFeedbacks =  await Task.Factory.StartNew(() => 
+                ValidateDataStream(stream), cancellationToken);
+
+            ResetValidator();
+
+            return validationFeedbacks;
+        }
+
+        private void SetValidationParameters(Encoding? _streamEncoding, PxFileSyntaxConf? _conf, int _numOfRows, int _rowLen, int _startRow)
+        {
+            streamEncoding = _streamEncoding ?? Encoding.Default;
+            conf  = _conf ?? PxFileSyntaxConf.Default;
+            expectedRows = _numOfRows;
+            expectedRowLength = _rowLen;
+            startRow = _startRow;
+
+            commonValidators.Add(new DataStructureValidator());
+            dataNumValidators.AddRange(commonValidators);
+            dataNumValidators.Add(new DataNumberValidator());
+            dataStringValidators.AddRange(commonValidators);
+            dataStringValidators.Add(new DataStringValidator());
+            dataSeparatorValidators.AddRange(commonValidators);
+            dataSeparatorValidators.Add(new DataSeparatorValidator());
+        }
+
+        private List<ValidationFeedback> ValidateDataStream(Stream stream)
+        {
+            List<ValidationFeedback> validationFeedbacks = [];
+            byte endOfData = (byte)conf.Symbols.EntrySeparator;
+            stringDelimeter = (byte)conf.Symbols.Value.StringDelimeter;
+            currentEntry = new(StreamBufferSize);
+            byte[] buffer = new byte[StreamBufferSize];
+            int bytesRead;
+            while ((bytesRead = stream.Read(buffer, 0, buffer.Length)) > 0)
+            {
```

Comment:
Then the async wrapper function calls this method, the state of the cancellation token should be checked before each read.